Automatic scaling

lindsayad opened the pull request:
Implements automatic scaling for both field and scalar variables.

In the current implementation we scale a variable's Jacobian and residual by a scalar corresponding to the maximum absolute value of the diagonal for that variable. In this way the Jacobian corresponding to each variable's physics is brought to order unity. 

If a variable does not have any diagonal dependence, such as for the pressure variable in the incompressible Navier Stokes equations, then its residuals/Jacobians will not be scaled. In the future it would be nice to add an enum for scaling options, similar to what is done for Jacobian preconditioning in [PETSc](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/PC/PCJACOBI.html).

Only kernels and scalar kernels currently contribute to the diagonal values for determining the scaling factors. Taking my comment from the code:
> Only compute Jacobians corresponding to the diagonals of volumetric compute objects because this typically gives us a good representation of the physics. NodalBCs and Constraints can introduce dramatically different scales (often order unity). IntegratedBCs and/or InterfaceKernels may use penalty factors. DGKernels may be ok, but they are almost always used in conjunction with Kernels.

Automatic scaling is not the default; if we make it the default, then we have many exodiffs that I do not feel like re-golding. I think we are planning on making non-automatic-scaling a legacy default. There are two ways to perform automatic scaling. The least expensive is to only compute the scaling factors a single time before the first time step. This would be achieved with this problem block:
```
[Problem]
  automatic_scaling = true
[]
```
The other way to do it as implemented in this PR is to compute a jacobian for scaling purposes at the beginning of every time step. This would be achieved with this block:
```
[Problem]
  automatic_scaling = true
  compute_scaling_once = false
[]
```
This is obviously more expensive. Probably a less expensive method as suggested by @friedmud would be to use scaling factors computed during a preceding time step during Jacobian calculations related to the actual non-linear solve. This would require caching the values for later application and would also require more code in `NonlinearSystemBase::computeJacobianInternal`. With the current implementation, we can just return from this function after computing the kernels and scalar kernels any time we know we are doing a "variable scaling jacobian." It would be a little tricky to interrupt a normal Jacobian computation being done through `ComputeJacobianThread` (we don't want penalty dirichlet BCs for instance) for the purpose of computing the future scaling factors.

lindsayad committed:
Initial automatic scaling commit

Refs #12601

lindsayad committed:
Debugging automatic scaling

Fix:
- Time derivatives for initial Jacobian computation. Now transient_simple_diffusion
  looks great
- Hanging node constraints introduce ones on the diagonal which messes with the scaling
  factor calculation. This messes up step 7 in the tutorial

lindsayad committed:
Add parameter for whether to perform scaling once

lindsayad committed:
Apply scaling factors to displaced non-linear system

lindsayad committed:
Only calculate for local dofs

lindsayad committed:
Don't do automatic scaling by default

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/349913/) on dacf0ac wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13547/site)

This comment will be updated on new commits.

tophmatthews commented:
Dude...you're like a genie...this is the second of my three wishes (mortar, scaling, and "more better" ref residual) @rcm59 @novasr

lindsayad commented:
This actually isn't doing anything right now...I'm going to add some tests

lindsayad commented:
You didn't have a wish for better Jacobians, e.g. AD 😆

tophmatthews commented:
AH CRAP! I knew i was forgetting something...okay, so four wishes. I'm needy.

lindsayad commented:
With automatic scaling, the convergence measure for the default convergence check (that provided by `FEProblem`) should actually be better

tophmatthews commented:
As in, the non-linear tolerances will be checked against a scaled residual? It will still be checked against the initial residual though correct? I think that motivated ref residual in part in order to carry across simulation time where not much is changing

lindsayad commented:
The scaling factors are calculated in an initial Jacobian evaluation that occurs before any residual calculations, so the initial residual will **include** the automatic scaling factors.

lindsayad committed:
Add automatic scaling tests

Demonstrates that we are not working in threaded mode yet

lindsayad committed:
Thread fix

tophmatthews commented on modules/doc/content/application_usage/failed_solves.md:
```diff
@@ -101,6 +101,25 @@ Initial residual before setting preset BCs: 65444.1
 
 In this case you are likely to require many nonlinear iterations as well, but the reason is that your linear iterations don't drop. This could be due to missing terms or errors in your Jacobian or because the way you are applying your preconditioner in PETSc is not good for the problem. Make sure your Jacobian is correct and add off-diagonal terms for multivariable problems.
 
+An additional possible reason for a poor linear solve is that your problem is very poorly
```
This is awesome info!!

lindsayad committed:
"White-list" objects for automatic scaling value calculations

Previously we were "black-listing" objects for this calculation. E.g. we would
not compute NodalBCs if we were computing the Jacobian for automatic scaling purposes.
Now we have white-listed kernels, e.g. when we're doing our scaling jacobian calculation
we only call through to kernels as opposed to *not* calling through to other things.
This makes the code for maintainable because now if a new type of compute object is added,
then the adder doesn't have to remember to prevent the scaling jacobian calculation from
calling through to it

lindsayad committed:
Make use of Communicator::max(std::vector)

lindsayad committed:
Add support for scaling scalar variables

tophmatthews commented on test/tests/kernels/simple_transient_diffusion/tests:
```diff
@@ -4,5 +4,73 @@
+    issues = '#12601'
+    design = 'failed_solves.md'
+    requirement = "We shall be able to solve an initially poorly scaled problem by using MOOSE's automatic scaling feature in parallel"
+    cli_args = 'Problem/automatic_scaling=true -pc_type hypre -pc_hypre_type boomeramg'
```
Hopefully this will work for superlu?

lindsayad replied:
Definitely

permcody reviewed (dismissed):
Looks good - nice work!

fdkong commented on framework/include/loops/ComputeJacobianForScalingThread.h:
```diff
@@ -0,0 +1,29 @@
+
+class FEProblemBase;
+typedef unsigned int TagID;
+
```
We have "TagID" defined in some header already. 

#include "MooseTypes.h"

lindsayad replied:
I generally like to avoid header dependencies when they are not needed. It can add significantly to compile times. I understand that this makes it harder to change the `typedef` in the future, but APIs shouldn't be that easy to change, should they? 😄

permcody replied:
Yeah - I'm not sure if having the same type defined in multiple places is a great idea even if it's with the best intentions. How about splitting up moose types so that we have a file without any includes in it (where that's possible) and then the rest in a different file? I'm open to other ideas here. I also don't think this is a showstopper.

lindsayad replied:
I've decided to just include `MooseTypes` here. That header really does a pretty good job of having minimal includes itself. It's just that people are throwing things in `MooseTypes` all the time so you'll have to recompile this a lot, but that's ok

fdkong commented on modules/doc/content/application_usage/failed_solves.md:
````diff
@@ -101,6 +101,25 @@ Initial residual before setting preset BCs: 65444.1
+
+```
+[Problem]
+  automatic_scaling = true
````
It might be better to add this parameter to executioner  block

permcody replied:
I'm not sure I agree. We use the `[Problem]` block for controlling a lot of application-level parameters already - even ones you might argue fit with the executioner just as well (e.g. `solve`). Also keep in mind that we will be adding the legacy flags so that we can control this behavior on an "App" to "App" basis. I think Problem is the right place for this.

fdkong replied:
I still like this kind of thing to Executioner. In order to use `auto scaling`, I have to create a problem block?

permcody replied:
Yes - but ideally users won't be doing this for an extended period of time. We'll get the legacy flag logic in there so individual apps can switch it on/off globally. Hopefully this isn't a parameter that you'll have to use very often if your application is making an effort to switch over to automatic scaling (regolding and evaluating failing tests).

lindsayad replied:
I am currently working on a commit where I've switched it over to the `Executioner`. I don't really care; it sounds like there are arguments both ways. In terms of code I think it fits much more naturally with the Problem, but for the user the Executioner seems better.

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3103,3 +3134,73 @@ NonlinearSystemBase::mortarJacobianConstraints(bool displaced)
+          if (dof_map.local_index(dof_index))
+          {
+            // For now we will use the diagonal for determining scaling
+            auto mat_value = petsc_matrix(dof_index, dof_index);
```
It might be expensive since we need to use MatGetRow and also need to search for column. 

It is better to grab the diagonal matrix.

Minor thing: we need to check if an index is locally owned

lindsayad replied:
Done

lindsayad committed:
Move scaling options to the Executioner

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3103,3 +3134,73 @@ NonlinearSystemBase::mortarJacobianConstraints(bool displaced)
+#ifdef LIBMESH_HAVE_PETSC
+  if (dynamic_cast<PetscMatrix<Real> *>(sys.matrix))
+  {
+    auto & petsc_matrix = *dynamic_cast<PetscMatrix<Real> *>(sys.matrix);
```
Done

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/351682/) on 05937bf : invalidated by @lindsayad

lindsayad committed:
Use MatGetDiagonal through libMesh interface to reduce expense

lindsayad committed:
Add timing

tophmatthews commented:
Go genie go!

![Culture_AladdinTrailer](https://user-images.githubusercontent.com/7014706/59457882-8d05dc00-8dd6-11e9-8268-273641ebf149.jpg)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/351839/) on 81d9f87 : invalidated by @permcody

Documentation build fixed

lindsayad committed:
Add App level option for whether to perform automatic scaling by default

lindsayad committed:
Turn off automatic_automatic_scaling in all repo apps

lindsayad committed:
Don't do automatic scaling by default

I think we need to vet automatic scaling a bit more before we make this
the default for future applications. For example we should at least be able
to make the MooseTest suite clean with auto scaling.

tophmatthews commented:
Hmmmm....I am getting a floating point error when running this with Newton, `line_search=none`, `petsc_options = '-snes_converged_reason -snes_linesearch_monitor'`and no other petsc_options. My input is too complex to share here, but here's the error:

```
Time Step 1, time = 1e-10, dt = 1e-10


Performing automatic scaling calculation

 0 Nonlinear |R| = [32m8.482386e+08[39m
      0 Linear |R| = [32m8.482386e+08[39m
      1 Linear |R| = [32m1.384385e-07[39m
 1 Nonlinear |R| = [32m2.686956e+07[39m
      0 Linear |R| = [32m2.686956e+07[39m

Floating point exception signaled (floating point divide by zero)!

To track this down, compile in debug mode, then in gdb do:
  break libmesh_handleFPE
  run ...
  bt
Stack frames: 38
0: 0   libmesh_opt.0.dylib                 0x000000010b62b09e libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 1198
1: 1   libmesh_opt.0.dylib                 0x000000010b626f8e libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 270
2: 2   libmesh_opt.0.dylib                 0x000000010b616936 (anonymous namespace)::libmesh_handleFPE(int, __siginfo*, void*) + 998
3: 3   libsystem_platform.dylib            0x00007fff7a694b5d _sigtramp + 29
4: 4   ???                                 0x00007fd207006800 0x0 + 140540037326848
5: 5   libmoose-opt.0.dylib                0x000000010abb2d63 SystemBase::reinitElem(libMesh::Elem const*, unsigned int) + 99
6: 6   libmoose-opt.0.dylib                0x000000010a8da877 FEProblemBase::reinitElem(libMesh::Elem const*, unsigned int) + 39
7: 7   libmoose-opt.0.dylib                0x000000010ab6dd24 ComputeResidualThread::onElement(libMesh::Elem const*) + 52
8: 8   libmoose-opt.0.dylib                0x000000010a85b454 ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) + 196
9: 9   libmoose-opt.0.dylib                0x000000010aba15be void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeResidualThread>(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, ComputeResidualThread&) + 62
10: 10  libmoose-opt.0.dylib                0x000000010ab9d772 NonlinearSystemBase::computeResidualInternal(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 898
11: 11  libmoose-opt.0.dylib                0x000000010ab9d1e1 NonlinearSystemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 321
12: 12  libmoose-opt.0.dylib                0x000000010a8f7299 FEProblemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 1017
13: 13  libmoose-opt.0.dylib                0x000000010a8f6a36 FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 118
14: 14  libmoose-opt.0.dylib                0x000000010a8f64d8 FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 56
15: 15  libmoose-opt.0.dylib                0x000000010ab8e059 ComputeResidualFunctor::residual(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, libMesh::NonlinearImplicitSystem&) + 41
16: 16  libmesh_opt.0.dylib                 0x000000010bb6a2af libmesh_petsc_snes_residual + 207
17: 17  libpetsc.3.9.dylib                  0x000000010de956cf SNESComputeFunction + 335
18: 18  libpetsc.3.9.dylib                  0x000000010dee2b8f SNESLineSearchApply_Basic + 415
19: 19  libpetsc.3.9.dylib                  0x000000010dee760f SNESLineSearchApply + 287
20: 20  libpetsc.3.9.dylib                  0x000000010deccb3d SNESSolve_NEWTONLS + 2125
21: 21  libpetsc.3.9.dylib                  0x000000010de9c560 SNESSolve + 1632
22: 22  libmesh_opt.0.dylib                 0x000000010bb69e18 libMesh::PetscNonlinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) + 888
23: 23  libmesh_opt.0.dylib                 0x000000010bbaff2b libMesh::NonlinearImplicitSystem::solve() + 283
24: 24  libmoose-opt.0.dylib                0x000000010ac633ef TimeIntegrator::solve() + 31
25: 25  libmoose-opt.0.dylib                0x000000010ab92547 NonlinearSystem::solve() + 503
26: 26  libmoose-opt.0.dylib                0x000000010a8f5161 FEProblemBase::solve() + 113
27: 27  libmoose-opt.0.dylib                0x000000010a6e4d26 FEProblemSolve::solve() + 22
28: 28  libmoose-opt.0.dylib                0x000000010a6e9d55 PicardSolve::solveStep(double, double&, double, double&, bool, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 1157
29: 29  libmoose-opt.0.dylib                0x000000010a6e88a2 PicardSolve::solve() + 2114
30: 30  libmoose-opt.0.dylib                0x000000010aad10c1 TimeStepper::step() + 33
31: 31  libmoose-opt.0.dylib                0x000000010a6efb68 Transient::takeStep(double) + 152
32: 32  libmoose-opt.0.dylib                0x000000010a6ef415 Transient::execute() + 85
33: 33  libmoose-opt.0.dylib                0x000000010aeaf4f0 MooseApp::executeExecutioner() + 96
34: 34  libmoose-opt.0.dylib                0x000000010aeafa13 MooseApp::run() + 227
35: 35  marmot-opt                          0x0000000109603b80 main + 128
36: 36  libdyld.dylib                       0x00007fff7a4a93d5 start + 1
37: 37  ???                                 0x0000000000000007 0x0 + 7
[0] /Users/topher/projects/moose/scripts/../libmesh/src/base/libmesh.C, line 134, compiled Jun 13 2019 at 12:34:56
We caught a libMesh error
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 1
[31m Solve Did NOT Converge![39m

Solve failed, cutting timestep.
```

tophmatthews commented:
...but It didn't fail when I had... 
```
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = ' hypre       boomeramg'
```

lu is default correct? should lu not be used @lindsayad ?

fdkong commented:
Could you make a small test case? Scaled system should allow to use whatever can be used in the unscaled system.

tophmatthews commented:
Yeah, let me try...

lindsayad commented:
Yea, it's possible there was a zero pivot or something in your linear solve. Make sure you also run with `-ksp_converged_reason` because it looks like you didn't get a first linear residual which tells me the linear solve failed. An initial residual of 1e8 is rather high; ordinarily the automatic scaling would help bring your residual towards unity...unless that high value is coming from something other than a kernel. I guess we'll learn more when you provide us with a test case.

tophmatthews commented:
Yeah, looks like a divide by zero, here it is with ksp_converged_reason
```
  Linear solve did not converge due to DIVERGED_PCSETUP_FAILED iterations 0
                 PCSETUP_FAILED due to FACTOR_NUMERIC_ZEROPIVOT 

Floating point exception signaled (floating point divide by zero)!
```
I'm running kernels with dampers. I'll try debug

tophmatthews commented:
```
Time Step 1, time = 1e-10, dt = 1e-10


Performing automatic scaling calculation

 0 Nonlinear |R| = 8.482386e+08
      0 Linear |R| = 8.482386e+08
      1 Linear |R| = 1.384385e-07
  Linear solve converged due to CONVERGED_RTOL iterations 1
 1 Nonlinear |R| = 2.686956e+07
      0 Linear |R| = 2.686956e+07
  Linear solve did not converge due to DIVERGED_PCSETUP_FAILED iterations 0
                 PCSETUP_FAILED due to FACTOR_NUMERIC_ZEROPIVOT 
Process 3537 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_ARITHMETIC (code=EXC_I386_SSEEXTERR, subcode=0x1f35)
    frame #0: 0x0000000103c75a10 libmoose-dbg.0.dylib`void libMesh::TypeVector<double>::add_scaled<double>(this=0x00000001482b17e8, p=0x00007ffeefbf92c8, factor=0x0000000149416e00) at type_vector.h:676
   673 	
   674 	#if LIBMESH_DIM == 3
   675 	  _coords[0] += factor*p(0);
-> 676 	  _coords[1] += factor*p(1);
   677 	  _coords[2] += factor*p(2);
   678 	#endif
   679 	
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_ARITHMETIC (code=EXC_I386_SSEEXTERR, subcode=0x1f35)
  * frame #0: 0x0000000103c75a10 libmoose-dbg.0.dylib`void libMesh::TypeVector<double>::add_scaled<double>(this=0x00000001482b17e8, p=0x00007ffeefbf92c8, factor=0x0000000149416e00) at type_vector.h:676
    frame #1: 0x0000000103c6853b libmoose-dbg.0.dylib`MooseVariableData<double>::computeValues(this=0x000000010f82ac00) at MooseVariableData.C:559
    frame #2: 0x0000000103c99df8 libmoose-dbg.0.dylib`MooseVariableFE<double>::computeElemValues(this=0x000000010c5815f0) at MooseVariableFE.C:365
    frame #3: 0x0000000104017cd9 libmoose-dbg.0.dylib`SystemBase::reinitElem(this=0x000000010f804218, (null)=0x000000010c5012b0, tid=0) at SystemBase.C:323
    frame #4: 0x0000000103556b14 libmoose-dbg.0.dylib`FEProblemBase::reinitElem(this=0x000000010f800418, elem=0x000000010c5012b0, tid=0) at FEProblemBase.C:1449
    frame #5: 0x0000000103efdbe8 libmoose-dbg.0.dylib`ComputeResidualThread::onElement(this=0x00007ffeefbfad50, elem=0x000000010c5012b0) at ComputeResidualThread.C:108
    frame #6: 0x00000001033ceb06 libmoose-dbg.0.dylib`ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator(this=0x00007ffeefbfad50, range=0x000000011f776a50, bypass_threading=false)(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) at ThreadedElementLoopBase.h:200
    frame #7: 0x0000000103fdc3f1 libmoose-dbg.0.dylib`void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeResidualThread>(range=0x000000011f776a50, body=0x00007ffeefbfad50) at threads_pthread.h:380
    frame #8: 0x0000000103fcfd17 libmoose-dbg.0.dylib`NonlinearSystemBase::computeResidualInternal(this=0x000000010f804218, tags=size=3) at NonlinearSystemBase.C:1343
    frame #9: 0x0000000103fcf549 libmoose-dbg.0.dylib`NonlinearSystemBase::computeResidualTags(this=0x000000010f804218, tags=size=3) at NonlinearSystemBase.C:674
    frame #10: 0x00000001035a6a17 libmoose-dbg.0.dylib`FEProblemBase::computeResidualTags(this=0x000000010f800418, tags=size=3) at FEProblemBase.C:4618
    frame #11: 0x00000001035a568b libmoose-dbg.0.dylib`FEProblemBase::computeResidualInternal(this=0x000000010f800418, soln=0x000000010c5aad60, residual=0x00007ffeefbfc8b0, tags=size=3) at FEProblemBase.C:4480
    frame #12: 0x00000001035a50ac libmoose-dbg.0.dylib`FEProblemBase::computeResidual(this=0x000000010f800418, soln=0x000000010c5aad60, residual=0x00007ffeefbfc8b0) at FEProblemBase.C:4435
    frame #13: 0x00000001035a4a12 libmoose-dbg.0.dylib`FEProblemBase::computeResidualSys(this=0x000000010f800418, (null)=0x000000010c5aa770, soln=0x000000010c5aad60, residual=0x00007ffeefbfc8b0) at FEProblemBase.C:4412
    frame #14: 0x0000000103fa22f2 libmoose-dbg.0.dylib`ComputeResidualFunctor::residual(this=0x000000010f8056f8, soln=0x000000010c5aad60, residual=0x00007ffeefbfc8b0, sys=0x000000010c5aa770) at ComputeResidualFunctor.C:23
    frame #15: 0x0000000107b8ec6f libmesh_dbg.0.dylib`::libmesh_petsc_snes_residual(snes=0x000000010f81ce20, x=0x0000000118e79620, r=0x000000010cae0620, ctx=0x000000010c5ab4d0) at petsc_nonlinear_solver.C:173
    frame #16: 0x000000010a5d26cf libpetsc.3.9.dylib`SNESComputeFunction + 335
    frame #17: 0x000000010a61fb8f libpetsc.3.9.dylib`SNESLineSearchApply_Basic + 415
    frame #18: 0x000000010a62460f libpetsc.3.9.dylib`SNESLineSearchApply + 287
    frame #19: 0x000000010a609b3d libpetsc.3.9.dylib`SNESSolve_NEWTONLS + 2125
    frame #20: 0x000000010a5d9560 libpetsc.3.9.dylib`SNESSolve + 1632
    frame #21: 0x0000000107b8d461 libmesh_dbg.0.dylib`libMesh::PetscNonlinearSolver<double>::solve(this=0x000000010c5ab4d0, pre_in=0x000000010c5ab400, x_in=0x000000010c5aac40, r_in=0x000000010c5ab220, (null)=0.01, (null)=1000) at petsc_nonlinear_solver.C:930
    frame #22: 0x0000000107ca85fe libmesh_dbg.0.dylib`libMesh::NonlinearImplicitSystem::solve(this=0x000000010c5aa770) at nonlinear_implicit_system.C:181
    frame #23: 0x000000010432f13d libmoose-dbg.0.dylib`TimeIntegrator::solve(this=0x000000010c5d5478) at TimeIntegrator.C:52
    frame #24: 0x0000000103fadb70 libmoose-dbg.0.dylib`NonlinearSystem::solve(this=0x000000010f804218) at NonlinearSystem.C:207
    frame #25: 0x000000010359fed8 libmoose-dbg.0.dylib`FEProblemBase::solve(this=0x000000010f800418) at FEProblemBase.C:4172
    frame #26: 0x0000000102ed6ed4 libmoose-dbg.0.dylib`FEProblemSolve::solve(this=0x000000010f80a308) at FEProblemSolve.C:139
    frame #27: 0x0000000102ee0faf libmoose-dbg.0.dylib`PicardSolve::solveStep(this=0x000000010f80a3e0, begin_norm_old=1.7976931348623157E+308, begin_norm=0x000000014805ba20, end_norm_old=1.7976931348623157E+308, end_norm=0x000000014803e110, relax=false, relaxed_dofs=size=0) at PicardSolve.C:378
    frame #28: 0x0000000102edec31 libmoose-dbg.0.dylib`PicardSolve::solve(this=0x000000010f80a3e0) at PicardSolve.C:199
    frame #29: 0x0000000103d0b684 libmoose-dbg.0.dylib`TimeStepper::step(this=0x00000001198209c8) at TimeStepper.C:156
    frame #30: 0x0000000102eee7fc libmoose-dbg.0.dylib`Transient::takeStep(this=0x000000010f80a218, input_dt=-1) at Transient.C:417
    frame #31: 0x0000000102eedf51 libmoose-dbg.0.dylib`Transient::execute(this=0x000000010f80a218) at Transient.C:311
    frame #32: 0x000000010484cceb libmoose-dbg.0.dylib`MooseApp::executeExecutioner(this=0x000000011800f218) at MooseApp.C:855
    frame #33: 0x000000010484e26f libmoose-dbg.0.dylib`MooseApp::run(this=0x000000011800f218) at MooseApp.C:965
    frame #34: 0x00000001000019d6 marmot-dbg`main(argc=7, argv=0x00007ffeefbfec38) at main.C:39
    frame #35: 0x00007fff7a4a93d5 libdyld.dylib`start + 1
    frame #36: 0x00007fff7a4a93d5 libdyld.dylib`start + 1
```

tophmatthews commented:
What does a zero pivot mean?

fdkong commented:
I think @lindsayad means you hit zero pivot in the matrix factorization using LU.  The LU does Gausian elelmemation, that is, divide by a factor.  Zero pivot means you the factor is zero. This happens often means your matrix is singular or nearly  singular or highly ill-conditioned.  Make the problem smaller, and then check the matrix.

lindsayad commented:
Of course our goal @fdkong is that through automatic scaling the user doesn’t have to worry about the condition number :-) but we’ll really need to see an input file to understand what’s going on. 

> On Jun 14, 2019, at 8:28 AM, Fande Kong <notifications@github.com> wrote:
> 
> I think @lindsayad means you hit zero pivot in the matrix factorization using LU. The LU does Gausian elelmemation, that is, divide by a factor. Zero pivot means you the factor is zero. This happens often means your matrix is singular or nearly singular or highly ill-conditioned. Make the problem smaller, and then check the matrix.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

tophmatthews commented:
ping @gardnerru
